packages/ai/src/workflows/controllers.ts

```diff
@@ -1,117 +1,165 @@
-import { PromptArtifact } from "../Artifact.js";
+import { CodeArtifact, PromptArtifact } from "../Artifact.js";
+import { LLMStepInstance } from "../LLMStepInstance.js";
 import { IOShape } from "../LLMStepTemplate.js";
 import { adjustToFeedbackStep } from "../steps/adjustToFeedbackStep.js";
 import { fixCodeUntilItRunsStep } from "../steps/fixCodeUntilItRunsStep.js";
 import { generateCodeStep } from "../steps/generateCodeStep.js";
 import { matchStyleGuideStep } from "../steps/matchStyleGuideStep.js";
+import { runAndFormatCodeStep } from "../steps/runAndFormatCodeStep.js";
 import { Workflow } from "./Workflow.js";
+import {
+  NextStepAction,
+  WorkflowGuardHelpers,
+} from "./WorkflowGuardHelpers.js";
 
-const MAX_RETRIES = 5;
-const MAX_ADJUSTS = 3;
-const MAX_STYLE_GUIDES = 2;
-const MAX_FIXES = 3;
+const MAX_MINOR_ERRORS = 5;
+
+type config = {
+  maxNumericSteps: number;
+  maxStyleGuideSteps: number;
+};
+
+// Error Messages
+const ERROR_MESSAGES = {
+  FAILED_STEP: (stepName: string, errorType: string) =>
+    `Step ${stepName} failed with error type ${errorType}`,
+  NOT_DONE: (stepName: string, state: string) =>
+    `Impossible, step ${stepName} is not done: ${state}`,
+  NO_CODE: "Impossible state, generate didn't return code",
+  UNKNOWN_STEP: "Unknown step",
+};
+
+function getNextIntendedState<Shape extends IOShape>(
+  h: WorkflowGuardHelpers<Shape>,
+  intendedStep: "AdjustToFeedback" | "MatchStyleGuide",
+  prompt: PromptArtifact,
+  code: CodeArtifact,
+  config: config
```

Comment:
You could have 2 parameters instead of 5 if you moved this function to be a closure inside `addLinearRule`

Reply:
fair, happy to move this.